### PR TITLE
Make trigger-based time-offset correction more generalisable (resample time before comparing it to trigger)

### PR DESCRIPTION
**Pull request by smoia**
Closes none, but addresses #306 

## Proposed Changes

  - `interfaces.acq` initialisation is now independent from the trigger
  - The trigger channel number is still passed to the file reading functions for objects initialisation
  - If time and trigger channel don't have the same length, time is (locally) resampled to match the trigger length

### CODE REVIEW

File: phys2bids/interfaces/acq.py
Comment (vinferrer) [inline phys2bids/interfaces/acq.py:43]:
```suggestion
    frequencies = [x.samples_per_second for x in data]
    max_index = [i for i, j in enumerate(frequencies) if j == max(frequencies)][0]
    freq = [data[max_index].samples_per_second, ]
    interval = 1/freq[0]
    duration = (timeseries[max_index].shape[max_index] + 1) * interval
    timeseries = [np.ogrid[0:duration:interval][:-1], ]  # create time channel
```
This will make your frequency time and time channel independent from any index because you will always get the biggest one

File: phys2bids/physio_obj.py
Comment (vinferrer) [inline phys2bids/physio_obj.py:464]: you wouldn't need this
```suggestion
```

File: phys2bids/physio_obj.py
Comment (vinferrer) [inline phys2bids/physio_obj.py:480]:
```suggestion
        time_offset = self.timeseries[0][timepoints.argmax()]
```

**Comment by smoia** [comment]
@vinferrer I understand your suggested changes, however please consider that:
- Even if you set your time to be taken from the channel with the highest frequency, you still need to check at a later point that its size matches with your trigger. It's not a given that your trigger channel is the one with the highest frequency. For instance, if for any reason a user decides that a PPG should be sampled at 1000 Hz, while they're happy with a 40 Hz trigger sampling, the suggested change would not provide the right time offset.
- The solution I'm proposing is very condensed in terms of code (in fact, it's two lines of code more that create an `if` statement, that might not be run). The suggested changes are more than twice the size, with two `for` loops. It would still need the original `if`statement (see point above). It's a lot of computation more for a feature I'm not sure is needed.

I wouldn't accept the suggested changes, but if a second reviewer (@eurunuela? @tsalo?) agrees on them, I'll change the code.

**Comment by vinferrer** [comment]
You have a good point there, I think we can go with your solution

**Review by vinferrer — APPROVED** [review]
LGTM!! however, we need one more reviewer

File: phys2bids/physio_obj.py
Comment (eurunuela) [inline phys2bids/physio_obj.py:464]: I believe these two lines here are not tested. Could you open an issue for this please? Or add a quick assert to the  test. Thanks!

**Review by eurunuela — COMMENTED** [review]
I'm happy with the changes, just want you to elaborate a bit on two of the lines as I may have missed the discussion here.

**Comment by smoia** [comment]
I added some asserts and a full test for this PR. I don't understand why codecov is unhappy though. @eurunuela let me know if there is something else I can do!

**Review by eurunuela — APPROVED** [review]
LGTM! Thank you @smoia !

**Comment by smoia** [comment]
@vinferrer if you're ok with the updates, you can approve and merge in (and then we start again with #306)!

File: phys2bids/interfaces/acq.py
Comment (tsalo) [inline phys2bids/interfaces/acq.py:15]: If `chtrig` isn't being used anymore (and based on the docstring change that's what I'm inferring), then why keep the argument?

File: phys2bids/physio_obj.py
Comment (tsalo) [inline phys2bids/physio_obj.py:515]: Why have this change?

File: phys2bids/tests/test_physio_obj.py
Comment (tsalo) [inline phys2bids/tests/test_physio_obj.py:94]: Why deepcopy this array?

File: phys2bids/physio_obj.py
Comment (tsalo) [inline phys2bids/physio_obj.py:465]: Is this worth a warning? Given how this interacts with #306, and depending on how uncommon this is, I think that having time not match the trigger might indicate that the automatic detection of the trigger channel made a mistake.

File: phys2bids/interfaces/acq.py
Comment (vinferrer) [reply phys2bids/interfaces/acq.py:15]: basically chtrig is called at the end of the function by ``populate_phys_auto``

Comment (tsalo) [reply phys2bids/interfaces/acq.py:15]: Ah, thank you. It needs to stay in the docstring then.

File: phys2bids/interfaces/acq.py
Comment (tsalo) [inline phys2bids/interfaces/acq.py:25]: Can you revert this? Perhaps something like

```python
    chtrig : int, optional
          index of trigger channel. Default is 0.
```

File: phys2bids/physio_obj.py
Comment (smoia) [reply phys2bids/physio_obj.py:515]: It's [this commit](https://github.com/physiopy/phys2bids/pull/308/commits/2d3e378b7f1ecf18b0ea482008ad3b9837bcc924). Basically, while testing I got some dtype errors of impossible subtraction between dtype X and dtype Y with some peculiar cases of `self.timeseries[0]` and `time_offset`. I read that an easy way to avoid them is using a complete subtraction statement rather than the previous version. It's a more global solution (despite possibly less elegant).

File: phys2bids/tests/test_physio_obj.py
Comment (smoia) [reply phys2bids/tests/test_physio_obj.py:94]: If we don't make a deep copy, it's just a shallow reference, so `test_time` changes when `blueprint_in.check_trigger_amount()` is called, and we wouldn't be able to set up the new asserts.

File: phys2bids/physio_obj.py
Comment (tsalo) [reply phys2bids/physio_obj.py:515]: That makes sense. I've had issues with in-place subtraction in the past too. Thanks for the explanation!

File: phys2bids/tests/test_physio_obj.py
Comment (tsalo) [reply phys2bids/tests/test_physio_obj.py:94]: I think that's a bug on `BlueprintInput`'s end rather than something that the test should work around.

Comment (vinferrer) [reply phys2bids/tests/test_physio_obj.py:94]: Actually this is related to the effect of shallow copies vs deepcopies: Basically, a shallow copy does copy the main object but references child objects to the original. Therefore if you modify the child object of the original, it also gets modified in the copy. If you deepcopy, you create a completely independent object (that consumes more memory). More info:
https://realpython.com/copying-python-objects/

Comment (vinferrer) [reply phys2bids/tests/test_physio_obj.py:94]: It is an undesired behavior but not a bug, just how python works

Comment (smoia) [reply phys2bids/tests/test_physio_obj.py:94]: Yup, @vinferrer is right. For our scopes, I don't think it's an undesired behaviour either.
However, if @tsalo you think there's a better solution to approach this, we can open an issue to change this - I wouldn't block this PR for this, though.

Comment (tsalo) [reply phys2bids/tests/test_physio_obj.py:94]: When I say it is a bug, I mean that a given function/class should not to modify its inputs directly unless that behavior is desired and documented. A good example of this is [`nilearn.image.get_data()`](https://nilearn.github.io/modules/generated/nilearn.image.get_data.html#nilearn.image.get_data), which is an alternative to nibabel's Nifti1Image's `get_fdata()`. Unlike nibabel's method, nilearn returns an img object's numpy array directly instead of a copy. Even with that warning, there are still unexpected behaviors like https://github.com/nilearn/nilearn/issues/2548, so I'd lean against it. That issue I linked was only a bug _after_ they switched from nibabel's function (return a copy) to nilearn's function (return a reference).

On the other hand, when the behavior isn't documented, we can get issues like https://github.com/nilearn/nilearn/issues/2281 and https://github.com/nilearn/nilearn/issues/2475. 

There are valid reasons to not make copies of all of your inputs within a function of course, like memory. I can't speak to the exact memory requirements of a BlueprintInput, but I imagine they could be quite large given the right circumstances. If you don't want to make a copy _within_ the BlueprintInput, the we should probably add something to the docstring about this behavior. `nilearn.image.get_data()`'s docstring could serve as a good template.

This is a minor thing, but since this numpy array is just an array of integers or floats, and not objects, shallow copying should be fine. You could use `test_time.copy()` without issue. The problem is that `BlueprintInput` doesn't make a copy of the array at all and operates on the original array directly.

Comment (smoia) [reply phys2bids/tests/test_physio_obj.py:94]: @tsalo now it's clear. Could you open an issue for this? I'm going to change the deepcopy to copy.

Comment (tsalo) [reply phys2bids/tests/test_physio_obj.py:94]: I've opened #314 (what luck!). I'll resolve this thread, but fixing the bug will involve dropping the copy in the test in a later PR. Thanks!

**Comment by vinferrer** [comment]
Should we merge then?

**Comment by vinferrer** [comment]
Or do you want us to re-review?

**Review by tsalo — APPROVED** [review]
LGTM.

**Pull request merged**
